### PR TITLE
Pushdown filters on window partitions

### LINKED ISSUE #10352: Partition predicate pushdown is not working with windows functions
### What happens?

The basetables sessions and treatments are delta tables partitioned by partition_year_month, the predicate pushdown works fine without any window function. But as I use a window function, although it contains the partition key itself, the pushdown does not work anymore, the filter is done after all data is aggregated. I just want to have the window function happening within the filtered partition, not across. How to change this behavior?

### To Reproduce

No window function, predicate pushdown works:

```
WITH
  m1 as (
    SELECT
      loguuid 
      ,sessionid                            
      ,partition_year_month
    FROM
     {{source('deltalake','sessions')}}
  )
  , m2 as (
    SELECT
      m1.*
    , treatmentstart
    FROM
      m1
    LEFT JOIN {{source('deltalake','treatments')}} tm 
    ON m1.sessionid = tm.sessionid and 
    m1.partition_year_month = tm.partition_year_month and 
    m1.loguuid = tm.loguuid 
  )
select *
from m2
```

image

Simple window function, predicate pushdown not working anymore:

```
WITH
  m1 as (
    SELECT
      loguuid 
      ,sessionid                            
      ,partition_year_month
    FROM
     {{source('deltalake','sessions')}}
  )
  , m2 as (
    SELECT
      m1.*
    , treatmentstart
    FROM
      m1
    LEFT JOIN {{source('deltalake','treatments')}} tm 
    ON m1.sessionid = tm.sessionid and 
    m1.partition_year_month = tm.partition_year_month and 
    m1.loguuid = tm.loguuid 
  )
select *
    , LAG(sessionid) OVER (PARTITION BY partition_year_month ORDER BY treatmentstart ASC) as prev_sessionid
from m2
```

image

### OS:

macos arm64

### DuckDB Version:

0.9.2

### DuckDB Client:

Python

### Full Name:

Norman Foerster

### Affiliation:

Fresenius Medical Care

### Have you tried this on the latest `main` branch?

I have not tested with any build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tmonster**
If a window has a partition, we can push it down. The logic being that if values in the partitioned column are omitted by the filter, then there is no need to calculate them in the window function/expression. 

Currently these filters can only be pushed down
- if every window expression in the window operator partitions by the filtered column
- The filter is not complex (i.e not a conjunction AND and not a function expression)

closes https://github.com/duckdb/duckdb/issues/10352

### CODE REVIEW

File: src/optimizer/pushdown/pushdown_window.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+			continue;
+		}
+		auto &window_expr = expr->Cast<BoundWindowExpression>();
+		auto &partitions = window_expr.partitions;
+		if (partitions.empty()) {
+			// all window expressions need to be partitioned by the same column
+			// in order to push down the window.
+			return FinishPushdown(std::move(op));
+		}
+		column_binding_set_t parition_bindings;
```
Comment (hawkfish) [inline src/optimizer/pushdown/pushdown_window.cpp:65]: "partition_bindings" (I do this one all the time!)

File: src/optimizer/pushdown/pushdown_window.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+				// if the filter has more columns, break to make sure the other
+				// columns of the filter are also partitioned on.
+				break;
+			}
+			// the filter is not on a column that is partitioned. immediately return.
+			filter_is_on_partition = false;
+			return;
+		}
+		// for simplicity, we don't push down filters that are functions.
+		// conjunction and filters we don't pushdown either
```
Comment (hawkfish) [inline src/optimizer/pushdown/pushdown_window.cpp:28]: AND

**Review by hawkfish — APPROVED** [review]
Very nice! Just some very minor nits.

**Review by hawkfish — APPROVED** [review]

**Review by hawkfish — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks! Looks good

**Pull request merged**